### PR TITLE
CLOUDSTACK-8822 - Replacing Runnable by Callable

### PR DESCRIPTION
**Pull request by wilderrodrigues**
That's the first part of the refactor, which will touch the ManagedContextRunnable and all its subclasses. However, in order to reduce impact, the first part comprises the Agent and Nio related classes (NioConnection, NioServer and NioClient).
 
* All the sub-classes were also updated according to the changes in the super-classes
* Improved exception handling
* There were also code formatting changes

Changes were structural and the NioTest covered them without need to modify the unit test.

This PR is quite extensive. Please, wait for the Test Report in order to proceed with further review.

ping @remibergsma @miguelaferreira @bhaisaab @karuturi @wido @DaanHoogland @borisroman @K0zka 

Cheers,
Wilder

### CODE REVIEW

File: engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java [hunk 1]
```diff
@@ -1315,6 +1321,10 @@ protected void doTask(final Task task) throws Exception {
                     } catch (final UnsupportedVersionException e) {
                         s_logger.warn(e.getMessage());
                         // upgradeAgent(task.getLink(), data, e.getReason());
+                    } catch (final ClassNotFoundException e) {
+                        final String message = String.format("Exception occured when executing taks! Error '%s'", e.getMessage());
```
Comment (bhaisaab) [inline engine/orchestration/src/com/cloud/agent/manager/AgentManagerImpl.java]: s/taks/task/ (minor issue)

**Comment by remibergsma** [comment]
Nice work @wilderrodrigues. Pinging @snuf to have a look.

**Comment by bhaisaab** [comment]
@wilderrodrigues LGTM (minor issue - s/taks/task)

**Comment by DaanHoogland** [comment]
pending pull-analysis results; LGTM

**Comment by wilderrodrigues** [comment]
Thanks, guys!

@bhaisaab: I will change that and push in the second part of the refactor.

Thanks again for a quick review! I will paste the tests results (still running because I'm testing a lot) before merging.

Cheers,
Wilder

**Comment by wilderrodrigues** [comment]
@remibergsma @bhaisaab @DaanHoogland 

Tests Results:

Environment

KVM + CentOS 7.1
Management Server + CentOS 7.1
Latest SytemVM
Agent + Common built from source


Test router internal advanced zone ... === TestName: test_02_router_internal_adv | Status : SUCCESS ===
ok
Test restart network ... === TestName: test_03_restart_network_cleanup | Status : SUCCESS ===
ok
Test router basic setup ... === TestName: test_05_router_basic | Status : SUCCESS ===
ok
Test router advanced setup ... === TestName: test_06_router_advanced | Status : SUCCESS ===
ok
Test stop router ... === TestName: test_07_stop_router | Status : SUCCESS ===
ok
Test start router ... === TestName: test_08_start_router | Status : SUCCESS ===
ok
Test reboot router ... === TestName: test_09_reboot_router | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 7 tests in 666.061s

OK
/tmp//MarvinLogs/test_routers_MT0LID/results.txt (END)



Test create VPC offering ... === TestName: test_01_create_vpc_offering | Status : SUCCESS ===
ok
Test VPC offering without load balancing service ... === TestName: test_03_vpc_off_without_lb | Status : SUCCESS ===
ok
Test VPC offering without static NAT service ... === TestName: test_04_vpc_off_without_static_nat | Status : SUCCESS ===
ok
Test VPC offering without port forwarding service ... === TestName: test_05_vpc_off_without_pf | Status : SUCCESS ===
ok
Test VPC offering with invalid services ... === TestName: test_06_vpc_off_invalid_services | Status : SUCCESS ===
ok
Test update VPC offering ... === TestName: test_07_update_vpc_off | Status : SUCCESS ===
ok
Test list VPC offering ... === TestName: test_08_list_vpc_off | Status : SUCCESS ===
ok
test_09_create_redundant_vpc_offering (integration.component.test_vpc_offerings.TestVPCOffering) ... === TestName: test_09_create_redundant_vpc_offering | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 8 tests in 1648.764s

OK
/tmp//MarvinLogs/test_vpc_offerings_JEYBTX/results.txt (END)



Test advanced zone virtual router ... === TestName: test_advZoneVirtualRouter | Status : SUCCESS ===
ok
Test Deploy Virtual Machine ... === TestName: test_deploy_vm | Status : SUCCESS ===
ok
Test Multiple Deploy Virtual Machine ... === TestName: test_deploy_vm_multiple | Status : SUCCESS ===
ok
Test Stop Virtual Machine ... === TestName: test_01_stop_vm | Status : SUCCESS ===
ok
Test Start Virtual Machine ... === TestName: test_02_start_vm | Status : SUCCESS ===
ok
Test Reboot Virtual Machine ... === TestName: test_03_reboot_vm | Status : SUCCESS ===
ok
Test destroy Virtual Machine ... === TestName: test_06_destroy_vm | Status : SUCCESS ===
ok
Test recover Virtual Machine ... === TestName: test_07_restore_vm | Status : SUCCESS ===
ok
Test migrate VM ... SKIP: At least two hosts should be present in the zone for migration
Test destroy(expunge) Virtual Machine ... === TestName: test_09_expunge_vm | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 10 tests in 1838.049s

OK (SKIP=1)
/tmp//MarvinLogs/test_vm_life_cycle_3JACRM/results.txt (END)



Test VPN in VPC ... === TestName: test_vpc_remote_access_vpn | Status : SUCCESS ===
ok
Test VPN in VPC ... === TestName: test_vpc_site2site_vpn | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 2 tests in 912.914s

OK
/tmp//MarvinLogs/test_vpc_vpn_ZJA1WV/results.txt (END)


VPN remote access user limit tests ... === TestName: test_01_VPN_user_limit | Status : SUCCESS ===
ok
Test create VPN when L2TP port in use ... === TestName: test_02_use_vpn_port | Status : SUCCESS ===
ok
Test create NAT rule when VPN when L2TP enabled ... === TestName: test_03_enable_vpn_use_port | Status : SUCCESS ===
ok
Test add new users to existing VPN ... === TestName: test_04_add_new_users | Status : SUCCESS ===
ok
Test add duplicate user to existing VPN ... === TestName: test_05_add_duplicate_user | Status : SUCCESS ===
ok
Test as global admin, add a new VPN user to an existing VPN entry ... === TestName: test_06_add_VPN_user_global_admin | Status : SUCCESS ===
ok
Test as domain admin, add a new VPN user to an existing VPN entry ... === TestName: test_07_add_VPN_user_domain_admin | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 7 tests in 2487.462s

OK
/tmp//MarvinLogs/test_vpn_users_RCQV0X/results.txt (END)


Test reset virtual machine on reboot ... === TestName: test_01_reset_vm_on_reboot | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 1 test in 654.363s

OK
/tmp//MarvinLogs/test_reset_vm_on_reboot_7NQTK2/results.txt (END)


Test create network in VPC ... === TestName: test_01_create_network_1_network_offering | Status : SUCCESS ===
ok
Test create network in VPC ... SKIP: Netscaler not configured: skipping test
Test create network in VPC mismatched services (Should fail) ... === TestName: test_02_create_network_fail_1_network_offering | Status : SUCCESS ===
ok
Test create network in VPC mismatched services (Should fail) ... SKIP: Netscaler not configured: skipping test
Test create multiple networks with LB service (Should fail) ... SKIP: Skipping test due to CLOUDSTACK-8437
Test create multiple networks with LB service (Should fail) ... SKIP: Skipping test due to CLOUDSTACK-8437
Test create network with redundant router capability ... SKIP: skipped - RvR didn't support VPC currently 
Test create network services not supported by VPC (Should fail) ... === TestName: test_07_create_network_unsupported_services | Status : SUCCESS ===
ok
Test create network without sourceNAT service in VPC (should fail) ... === TestName: test_08_create_network_without_sourceNAT | Status : SUCCESS ===
ok
Test create network with shared network offering ... === TestName: test_09_create_network_shared_nwoff_1_network_off_shared | Status : SUCCESS ===
ok
Test create network with shared network offering ... SKIP: Netscaler not configured: skipping test
Test create network with conserve mode ON ... === TestName: test_10_create_network_with_conserve_mode_1_network_offering | Status : SUCCESS ===
ok
Test create network with conserve mode ON ... === TestName: test_10_create_network_with_conserve_mode_2_network_offering_vpcns | Status : SUCCESS ===
ok
Test create network outside cidr range of VPC ... === TestName: test_01_create_network_outside_range_1_network_offering | Status : SUCCESS ===
ok
Test create network outside cidr range of VPC ... SKIP: Netscaler not configured: skipping test
Test create network outside cidr range of VPC ... === TestName: test_02_create_network_outside_range | Status : SUCCESS ===
ok
Test create network inside cidr range of VPC ... === TestName: test_03_create_network_inside_range_1_network_offering | Status : SUCCESS ===
ok
Test create network inside cidr range of VPC ... SKIP: Netscaler not configured: skipping test
Test create network overlapping cidr range of VPC ... === TestName: test_04_create_network_overlapping_range_1_network_offering | Status : SUCCESS ===
ok
Test create network overlapping cidr range of VPC ... SKIP: Netscaler not configured: skipping test
Test create network from different account in VPC ... === TestName: test_05_create_network_diff_account_1_network_offering | Status : SUCCESS ===
ok
Test create network from different account in VPC ... SKIP: Netscaler not configured: skipping test

----------------------------------------------------------------------
Ran 22 tests in 2202.933s

OK (SKIP=10)
/tmp//MarvinLogs/test_vpc_network_Z01VMJ/results.txt (END)

**Pull request merged**